File: Assets/Plugins/Android/src/com/playphone/multinet/unity/MNPlayerListProviderUnity.java
package com.playphone.multinet.unity;

import com.playphone.multinet.*;
import com.unity3d.player.UnityPlayer;

public class MNPlayerListProviderUnity {
    public static void shutdown() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getPlayerListProvider().shutdown();
            }
        });
    }

    public static String getPlayerList() {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getPlayerListProvider().getPlayerList());
    }

    protected static class MNPlayerListProviderEventHandler implements com.playphone.multinet.providers.MNPlayerListProvider.IEventHandler {

        @Override
        public void onPlayerJoin(final MNUserInfo player) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onPlayerJoin",player);
                }
            });
        }

        @Override
        public void onPlayerLeft(final MNUserInfo player) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onPlayerLeft",player);
                }
            });
        }
    }

    private static MNPlayerListProviderEventHandler registeredEventHandler = null;
    public static synchronized boolean registerEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getPlayerListProvider() == null) {
            MNUnity.ELog("MNPlayerListProvider is not ready. Use MNDirect\'s sessionReady event for adding your delegates.");
            return false;
        }

        if (registeredEventHandler == null) {
            registeredEventHandler = new MNPlayerListProviderUnity.MNPlayerListProviderEventHandler();
            MNDirect.getPlayerListProvider().addEventHandler(registeredEventHandler);
        }

        return true;
    }
    
    public static synchronized boolean unregisterEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getPlayerListProvider() == null) {
            return false;
        }

        if (registeredEventHandler != null) {
            MNDirect.getPlayerListProvider().removeEventHandler(registeredEventHandler);
            registeredEventHandler = null;
        }

        return true;
    }
}

File: Assets/Plugins/iOS/PlayPhoneMultiNet/unity/providers/MNPlayerListProviderUnity.m
//
//  MNPlayerListProviderUnity.m
//  PlayPhone SDK for Unity
//
//  Copyright (c) 2012 PlayPhone Inc. All rights reserved.
//

#import "MNDirect.h"
#import "MNPlayerListProvider.h"

#import "MNPlayerListProviderUnity.h"

@interface MNPlayerListProviderUnity : NSObject <MNPlayerListProviderDelegate>
+ (id)shared;
- (void)onPlayerJoin:(MNUserInfo*)player;
- (void)onPlayerLeft:(MNUserInfo*)player;
@end

static MNPlayerListProviderUnity *MNPlayerListProviderUnityInstance = nil;

@implementation MNPlayerListProviderUnity

+(id)shared {
    if (MNPlayerListProviderUnityInstance == nil) {
        MNPlayerListProviderUnityInstance = [[MNPlayerListProviderUnity alloc]init];
    }
    
    return MNPlayerListProviderUnityInstance;
}

- (void)onPlayerJoin:(MNUserInfo*)player {
    [MNUnity callUnityFunction:@"MNUM_onPlayerJoin" withParams:(player == nil ? [NSNull null] : player), nil];
}
- (void)onPlayerLeft:(MNUserInfo*)player {
    [MNUnity callUnityFunction:@"MNUM_onPlayerLeft" withParams:(player == nil ? [NSNull null] : player), nil];
}
@end

EXTERN_C
const char* _MNPlayerListProvider_GetPlayerList() {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect playerListProvider] getPlayerList]] UTF8String]);
}

EXTERN_C
bool _MNPlayerListProvider_RegisterEventHandler(){
    MARK;

    if ([MNDirect playerListProvider] == nil) {
        return false;
    }

    [[MNDirect playerListProvider] addDelegate:[MNPlayerListProviderUnity shared]];

    return true;
}

EXTERN_C
bool _MNPlayerListProvider_UnregisterEventHandler(){
    MARK;

    if ([MNDirect playerListProvider] == nil) {
        return false;
    }

    [[MNDirect playerListProvider] removeDelegate:[MNPlayerListProviderUnity shared]];

    return true;
}


